src/gls/ring_buffer.c3

```diff
@@ -0,0 +1,156 @@
+@require_once "std/wrappers.c3"
+@require_once "std/span.c3"
+
+// See https://lo.calho.st/posts/black-magic-buffer/ for an explanation.
+typedef RingBuffer : {
+    buffer : Span<u8>,
+    size : isize,
+    head : isize,
+    tail : isize,
+}
+
+function make<RingBuffer> : (
+    size : isize, name : CString
+) -> Optional<RingBuffer> = {
+    // TODO get the page size from sysconf().
+    if size % 4096 != 0 {
+        return make<Optional<RingBuffer>>();
+    }
+
+    let fd : int = memfd_create(name, 0x0);
+    if fd < 0 {
+        return make<Optional<RingBuffer>>();
+    }
+
+    if ftruncate(fd, size) != 0 {
+        sys_close(fd);
+        return make<Optional<RingBuffer>>();
+    }
+
+    let addr : iptr = mmap(
+        0, 2 * size, SYS_PROT_NONE(), SYS_MAP_PRIVATE() | SYS_MAP_ANONYMOUS(),
+        -1, 0
+    );
+
+    mmap(
+        addr, size, SYS_PROT_READ() | SYS_PROT_WRITE(),
+        SYS_MAP_SHARED() | SYS_MAP_FIXED(), fd, 0
+    );
+    mmap(
+        addr + size, size, SYS_PROT_READ() | SYS_PROT_WRITE(),
+        SYS_MAP_SHARED() | SYS_MAP_FIXED(), fd, 0
+    );
+
+    sys_close(fd);
+
+    return make<Optional<RingBuffer>>({
+        .buffer = { &mut addr_to_mut_heap_array<u8>(addr), 2 * size },
+        .size = size,
+        .head = 0,
+        .tail = 0,
+    });
+}
+
+function deallocate : (self : RingBuffer mut&) -> void = {
+    let addr : iptr = ref_to_addr(&self.buffer.data);
+
+    self:clear();
+
+    runtime_assert(munmap(addr, 2 * self.size) == 0);
+
+    let null : iptr = 0;
+    self.buffer = { &mut addr_to_mut_heap_array<u8>(null), 0 };
+    self.size = 0;
+}
+
+function read_impl : (
+    buffer : RingBuffer mut&, source : File
+) -> isize = {
+    // We need a span over the free space, so it's easier if we just unwrap the
+    // head.
+    mut head : isize = buffer.head;
+    if head <= buffer.tail {
+        head += buffer.size;
+    }
+
+    let free : Span<u8> = buffer.buffer:slice_between(buffer.tail, head - 1);
+    let view : View<u8> = source:read(free);
+
+    buffer.tail += view:length();
+    buffer.tail %= buffer.size;
+
+    return view:length();
+}
+
+function find_newline : (buffer : RingBuffer mut&) -> Optional<isize> = {
+    mut tail : isize = buffer.tail;
+    if tail < buffer.head {
+        tail += buffer.size;
+    }
+
+    for i in range(buffer.head, tail) {
+        if buffer.buffer:get(i) == char("\n") {
+            return make<Optional<isize>>(i);
+        }
+    }
+
+    return make<Optional<isize>>();
+}
+
+function read_line : (
+    buffer : RingBuffer mut&, source : File
+) -> Optional<View<u8>> = {
+    while true {
+        // Find the first '\n'.
+        let newline_idx : Optional<isize> = buffer:find_newline();
+        if newline_idx:has_value() {
+            return make<Optional<View<u8>>>(to_view(buffer.buffer:slice_between(buffer.head, newline_idx:data() + 1)));
+        }
+
+        // Read from the fd and try again.
+        if buffer:read_impl(source) == 0 {
```

Comment:
I'm assuming that this is racey.... Will this cause the server to break if the sender is just being slow between chunks?

Reply:
Hmm, why racey? We don't use `O_NONBLOCK` so the `read` syscall in there should block until data is available.

Reply:
Ahh, ofc that makes sense. I take it all back.